getnacelle/nacelle-react
# [packages/react-hooks: 4.0.0] Major Refactor to Fix Bugs & Improve Flexibility

### Stories
- [DD-352](https://nacelle.atlassian.net/browse/DD-352)

> As a merchant developer, I want line items created by @nacelle/react-hooks's useCart hook to expose more product data, so that I have access to data submitted to analytics services

- [DD-316](https://nacelle.atlassian.net/browse/DD-316)

> @nacelle/react-hooks is missing a getCheckout call to correctly set the checkoutComplete value

- [DD-307](https://docs.getnacelle.com/integrations/sanity-webhooks.html)

> As a merchant developer, I want the useCart hook in @nacelle/react-hooks to leverage inversion of control for cart actions, so that I can implement my own custom cart logic in my React storefront without needing to create an entire custom cart solution

### Type of Change

- [x] Breaking feature (requires bump to next major version, `4.0.0`)

### What is being changed and why?



NOTE: The updates to `useCart` and `useCheckout` are documented in the [README](https://github.com/getnacelle/nacelle-react/blob/DD-352/packages/react-hooks/README.md) - please check it out!

#### Persisting Checkout Data

Previously, the `checkoutComplete` and `checkoutId` values were properties of `cartState`, which is a bit confusing. Worse, the `checkoutComplete` and `checkoutId` were not persisted after visiting the checkout URL. This, coupled with the lack of a `getCheckout` call in the `useCheckout` hook, made it impossible to know whether or not a checkout had been completed or not.

Now, there is a new `checkoutState` which is made available to the entire app via a `<CheckoutProvider />`, which mirrors the Context Provider > `[state, actions]` (via `useReducer`) pattern used by the `useCart` hook. The `checkoutState` is persisted in Local Storage, and its values are used to perform a `getCheckout` call when needed, to check for a change in the `checkoutState.checkoutComplete` value.

##### Demo

###### Before Checkout

<img width="603" alt="before-checkout" src="https://user-images.githubusercontent.com/5732000/117330716-f926d880-ae63-11eb-8d74-b0b7a69d7b50.png">

###### After Incomplete Checkout

<img width="603" alt="after-incomplete-checkout" src="https://user-images.githubusercontent.com/5732000/117330750-0348d700-ae64-11eb-8d79-48f9772e26a6.png">

###### After Complete Checkout

<img width="603" alt="after-complete-checkout" src="https://user-images.githubusercontent.com/5732000/117330820-13f94d00-ae64-11eb-9aea-19050c287842.png">

#### More Flexible Types

The `CartItem` type forced the merchant to use a subset of `product` data, and colocated that data with some of the `variant` data. This could create confusion, and made it difficult for the merchant developer to get `product` properties which weren't included in `CartItem`.

The new `CartItem` type has three properties: `{ product, variant, quantity }`. This provides the merchant developer with added control over the contents of a `CartItem`, and provides a clear delineation of `product` and `variant` data.

#### Injectable Cart Logic

Merchant developers sometimes need to use logic that is specific to their project, to drive certain cart functionality. 

Previously, the merchant developer was required to use the default `useCart` reducer actions. 

Now, the `<CartProvider />` component can be provided with user-defined actions which replace the default actions (`addToCart`, `incrementItem`, etc.).

### How to Test



From the monorepo root:
- `npm run bootstrap`
- `cd packages/react-hooks`
- `npm run test:ci`

From the `nextjs` or `gatsby` example projects:
- in `components/Cart/Cart.js`, remove the `useEffect` block that clears the checkout and cart data when checkout is complete:
```jsx
// Remove This!
useEffect(() => {
  if (checkoutComplete) {
    clearCheckoutData();
    cartActions.clearCart();
  }
}, [checkoutComplete, clearCheckoutData, cartActions]);
```
- `npm run dev`
- add items to cart
- adjust item quantities
- perform a partial checkout
- return to the app, and use the [React Devtools](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en) to inspect components which use the `useCart` and `useCheckout` hooks (e.g. the `Cart` component in `examples/nextjs`)
- checkout again, and complete the checkout via the bogus payment gateway (this is set up for Nacelle spaces connected to the Prairie Wind Apparel Shopify store)...hmu in Slack for further instructions, if needed
- examine the `checkoutState` returned by the `useCheckout` hook - the `checkoutComplete` value should be set to `true`
